src/UCTNode.cpp

```diff
@@ -377,64 +441,96 @@ class NodeComp : public std::binary_function<UCTNode::node_ptr_t&,
 
 void UCTNode::sort_children(int color) {
     LOCK(get_mutex(), lock);
-    std::stable_sort(rbegin(m_children), rend(m_children), NodeComp(color));
+
+    // Because sort depends on visits there's no need to expand all nodes.
+    assert(!m_expanded.empty());
+    std::stable_sort(rbegin(m_expanded), rend(m_expanded), NodeComp(color));
 }
 
 UCTNode& UCTNode::get_best_root_child(int color) {
     LOCK(get_mutex(), lock);
-    assert(!m_children.empty());
 
-    return *(std::max_element(begin(m_children), end(m_children),
+    // Check if any of m_expanded have visits.
```

Comment:
Can this loop be avoided by doing something like:
if expanded is empty return expanded best child
best_expanded = max(expanded)
if best_expanded visits == 0 return expanded best child
return best_expanded
?

Reply:
Almost works,

when best_expanded.visits == 0 && best_expanded.score > best_unexpanded.score current code would return best_expanded, your pseudo code would return best_unexpanded

Reply:
Whoops, so it does. In that case the best_expanded and best_unexpanded scores can be compared to decide which to return.

Reply:
I think that would make the score sufficiently more complex (I'd either have to have a duplicate code of NodeComp logic or expand(best_child_score) either temporarily with logic to handle that or permanently in which case it's the same as this code.

Reply:
Agreed. Oh well.
Maybe I’ll have a play to see if there is any worthwhile performance gain by not expanding the best unexpanded to warrant e.g. making a function to compare an unexpanded to an expanded node that calls a function with the common code from NodeComp pulled out.
Shall I remove this review comment?